File: src/stores/knowledge.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import { useTauriIPC } from "../composables/useTauriIPC";

export interface KnowledgeDocument {
  id: string;
  project_id: string;
  title: string;
  content: string;
  file_path: string | null;
  doc_type: string;
  chunk_count: number;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeSearchResult {
  doc_id: string;
  title: string;
  snippet: string;
  source_type: string;
  rank: number;
}

export const useKnowledgeStore = defineStore("knowledge", () => {
  const { call } = useTauriIPC();
  const documents = ref<KnowledgeDocument[]>([]);
  const searchResults = ref<KnowledgeSearchResult[]>([]);
  const loading = ref(false);
  const searching = ref(false);

  const fetchDocuments = async (projectId: string) => {
    loading.value = true;
    try {
      documents.value = await call<KnowledgeDocument[]>("list_knowledge", {
        projectId,
      });
    } finally {
      loading.value = false;
    }
  };

  const importDocument = async (
    projectId: string,
    title: string,
    content: string,
    filePath?: string
  ) => {
    const doc = await call<KnowledgeDocument>("import_knowledge", {
      projectId,
      title,
      content,
      filePath: filePath || null,
    });
    documents.value.unshift(doc);
    return doc;
  };

  const deleteDocument = async (id: string) => {
    await call("delete_knowledge", { id });
    documents.value = documents.value.filter((d) => d.id !== id);
  };

  const search = async (projectId: string, query: string, limit?: number) => {
    if (!query.trim()) {
      searchResults.value = [];
      return;
    }
    searching.value = true;
    try {
      searchResults.value = await call<KnowledgeSearchResult[]>(
        "search_knowledge",
        {
          projectId,
          query,
          limit: limit || 20,
        }
      );
    } finally {
      searching.value = false;
    }
  };

  const clearSearch = () => {
    searchResults.value = [];
  };

  return {
    documents,
    searchResults,
    loading,
    searching,
    fetchDocuments,
    importDocument,
    deleteDocument,
    search,
    clearSearch,
  };
});

File: src-tauri/src/commands/knowledge.rs
use crate::db::models::knowledge::{CreateKnowledgeDocumentRequest, KnowledgeDocument};
use crate::error::AppResult;
use crate::state::AppState;
use serde::Serialize;
use tauri::State;

#[derive(Debug, Serialize, sqlx::FromRow)]
pub struct KnowledgeSearchResult {
    pub doc_id: String,
    pub title: String,
    pub snippet: String,
    pub source_type: String,
    pub rank: f64,
}

#[tauri::command]
pub async fn import_knowledge(
    state: State<'_, AppState>,
    project_id: String,
    title: String,
    content: String,
    file_path: Option<String>,
) -> AppResult<KnowledgeDocument> {
    let db = state.db().await?;
    let req = CreateKnowledgeDocumentRequest {
        project_id,
        title,
        content,
        file_path,
        doc_type: None,
    };
    let doc = KnowledgeDocument::create(&db, &req).await?;
    Ok(doc)
}

#[tauri::command]
pub async fn list_knowledge(
    state: State<'_, AppState>,
    project_id: String,
) -> AppResult<Vec<KnowledgeDocument>> {
    let db = state.db().await?;
    let docs = KnowledgeDocument::list_by_project(&db, &project_id).await?;
    Ok(docs)
}

#[tauri::command]
pub async fn delete_knowledge(
    state: State<'_, AppState>,
    id: String,
) -> AppResult<()> {
    let db = state.db().await?;
    KnowledgeDocument::delete(&db, &id).await?;
    Ok(())
}

#[tauri::command]
pub async fn search_knowledge(
    state: State<'_, AppState>,
    project_id: String,
    query: String,
    limit: Option<i32>,
) -> AppResult<Vec<KnowledgeSearchResult>> {
    let db = state.db().await?;
    let limit = limit.unwrap_or(20);

    let results = sqlx::query_as::<_, KnowledgeSearchResult>(
        r#"SELECT
            k.id as doc_id,
            k.title,
            snippet(knowledge_documents_fts, 1, '<mark>', '</mark>', '...', 32) as snippet,
            'knowledge' as source_type,
            rank
        FROM knowledge_documents_fts
        JOIN knowledge_documents k ON k.rowid = knowledge_documents_fts.rowid
        WHERE knowledge_documents_fts MATCH ? AND k.project_id = ?
        UNION ALL
        SELECT
            c.id as doc_id,
            c.title,
            snippet(chapters_fts, 2, '<mark>', '</mark>', '...', 32) as snippet,
            'chapter' as source_type,
            rank
        FROM chapters_fts
        JOIN chapters c ON c.id = chapters_fts.chapter_id
        WHERE chapters_fts MATCH ? AND c.project_id = ?
        ORDER BY rank
        LIMIT ?"#,
    )
    .bind(&query)
    .bind(&project_id)
    .bind(&query)
    .bind(&project_id)
    .bind(limit)
    .fetch_all(&db)
    .await?;

    Ok(results)
}
